### PR TITLE
Apply WELPI CTF Scaling at Schedule Level

### PR DESCRIPTION
**Pull request by bska**
This PR adds a new member function
```C++
Schedule::applyWellProdIndexScaling(well_name, report_step, scalingFactor)
```
which applies WELPI-based CTF scaling (by `scalingFactor`) for all pertinent report steps from `report_step` and until the end of the simulation run.  We take care to apply the relevant scaling exactly once, and in order to guarantee this we use ~~`OrderedMap::unique()`~~ `DynamicState<>::unique()` and also add a special-purpose predicate
```C++
Well::hasSameConnectionsPointers(const Well& other)
```
which returns true if and only if the internal `WellConnections` pointers of `*this` and `other` are the same&mdash;meaning `*this` and `other` share the exact same `WellConnections` object.

We also add a communications channel which determines the connections' WELPI-based CTF scaling eligibility in the time direction.  This is needed to ensure we don't apply scaling to CTFs which are added/modified **after** the WELPI-based scaling is logically applied (e.g., new `COMPDAT` records **after** `WELPI`).

Clients are expected to run a call to `Schedule::applyWellProdIndexScaling()` when triggered by a [`WELL_PRODUCTIVITY_INDEX`](https://github.com/OPM/opm-common/blob/641b9fb4b18196bd5b745b5f3e8cf25eab7a9975/opm/parser/eclipse/EclipseState/Schedule/Events.hpp#L105-L108) event that combines a desired PI value for a well with a calculation of the current effective PI value.  To that end, also split `Well`'s `applyWellProdIndexScaling()` function into two parts, one which calculates the scaling factor and one which actually scales the connections' CTF value.  The second of these is used in the implementation of the new `Schedule`-level member function.

Finally, adding this member function allows us to remove the poorly designed [`WELL_CONNECTIONS_UPDATED`](https://github.com/OPM/opm-common/blob/641b9fb4b18196bd5b745b5f3e8cf25eab7a9975/opm/parser/eclipse/EclipseState/Schedule/Events.hpp#L110-L114) event.  We don't actually need that event when we have the new member function.

---

Note that this PR depends on #2018 and must not be merged before that PR is merged into the master branch.

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
I've rebased this onto the current master branch which contains #2018 that was merged earlier.  I'm marking this as "ready for review" now.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
>[...] and in order to guarantee this we use `OrderedMap::unique()`

I guess you mean `DynamicState<T>::unique()`?

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp [hunk 1]
```diff
@@ -195,10 +195,17 @@ inline std::array< size_t, 3> directionIndices(const Opm::Connection::Direction
         return update;
     }
 
-    void WellConnections::applyWellPIScaling(const double scaleFactor)
+    void WellConnections::applyWellPIScaling(const double       scaleFactor,
+                                             std::vector<bool>& scalingApplicable)
     {
-        for (auto& conn : this->m_connections)
-            conn.applyWellPIScaling(scaleFactor);
+        scalingApplicable.resize(std::max(scalingApplicable.size(), this->m_connections.size()), true);
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp:201]: The `scalingApplicable`vector seems unused to me - always `true` for all elements?

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 2]
```diff
@@ -1499,6 +1499,37 @@ class ScheduleLogger {
         }
     }
 
+    void Schedule::applyWellProdIndexScaling(const std::string& well_name, const std::size_t reportStep, const double scalingFactor) {
+        auto wstat = this->wells_static.find(well_name);
+        if (wstat == this->wells_static.end())
+            return;
+
+        auto uwell = wstat->second.unique();
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:1507]: SInce this is borderline complex the `uwell` variable could maybe be `unique_wells`?

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 3]
```diff
@@ -1499,6 +1499,37 @@ class ScheduleLogger {
...
+            //     time                 < reportStep
+            return time_well_pair.first < lookup;
+        });
+
+        if (start == end)
+            // Report step after last?
+            return;
+
+        // Relies on wells_static being OrderedMap<string, DynamicState<shared_ptr<>>>
+        // which means uwell is a vector<pair<time, shared_ptr<>>>
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:1521]: Maybe rather `vector<std::size_t, shared_ptr<>>` - now one could think time was e.g. `std::time_t` or something similar?

**Comment by joakim-hove** [comment]
This is good; you taken quite complex interactions into limited amount of code! One question about the `applicableForScvaling`which seemed superfluos to me - otherwise GO!

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp (hunk 1 above)
Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp:201]: It is an in/out parameter.  The overall process is somewhat subtle however and I'd very much like a better suggestion.

The `resize()` call (not `assign()`) is to handle new connections being added during the time series, and all connections are **initially** eligible for `WELPI` scaling.  This also means the caller, currently the new member function `Schedule::applyWellProdIndexScaling()` only, can start by passing an empty `scalingApplicable` array.  We then possibly turn off the eligibility status in the statement
```C++
scalingApplicable[i] = conn.applyWellScaling(scaleFactor);
```
at which point it remains turned off for the remainder of the time series assuming the caller passes the same `scalingApplicable` array into the next call to `WellConnections::applyWellPIScaling()`.

In other words, the `scalingApplicable` provides a way to communicate eligibility status between the unique `WellConnections` instances.

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 3 above)
Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:1521]: > Maybe rather `vector<std::size_t, shared_ptr<>>` - now one could think time was e.g. `std::time_t` or something similar?

I'm on the fence here.  The **value** is a time (a report step), but the **type** is obviously a `std::size_t`.  How about
```C++
vector<pair<report_step, shared_ptr<>>>
```
instead?

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 2 above)
Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:1507]: > Since this is borderline complex the `uwell` variable could maybe be `unique_wells`?

Good idea.  I'll update accordingly.

**Comment by bska** [comment]
> > [...] and in order to guarantee this we use `OrderedMap::unique()`
> 
> I guess you mean `DynamicState<T>::unique()`?

You're right.  I'll update the description.

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 3 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:1521]: I like the alternative with `report_step`

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp:201]: Ok - I'll have another look, but to me it looked like there was never an assignment to `false`?

Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp:201]: > it looked like there was never an assignment to `false`?

I suppose this too is a bit subtle.  [`Connection::applyWellPIScaling()`](https://github.com/OPM/opm-common/blob/91bb168ee58d40875c0f2598d4af205d50166767/src/opm/parser/eclipse/EclipseState/Schedule/Well/Connection.cpp#L259-L265) returns `false` when the connection is not eligible for WELPI scaling.  In other words, the [update loop](https://github.com/OPM/opm-common/blob/91bb168ee58d40875c0f2598d4af205d50166767/src/opm/parser/eclipse/EclipseState/Schedule/Well/WellConnections.cpp#L204-L208) also assigns the `false` values if needed:
```C++
for (auto& conn : this->m_connections) {
    if (scalingApplicable[i])
        scalingApplicable[i] = conn.applyWellPIScaling(scaleFactor);
    ++i;
}
```

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 2 above)
Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: > Since this is borderline complex the `uwell` variable could maybe be `unique_wells`?

I've renamed the variable to `unique_well_instances`.

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 3 above)
Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: > I like the alternative with `report_step`

I've updated the comment accordingly.

**Comment by bska** [comment]
I've pushed an update to address the review comments.  I'll run a new build check.

**Comment by bska** [comment]
jenkins build this please
